### PR TITLE
[naga wgsl-in] Clamp negative f16 values to 0 when converting to unsigned integers during const evaluation

### LINKED ISSUE #7420: [naga wgsl-in] Panic when converting negative f16 to u32 during const evaluation
**Description**
We panic [here](https://github.com/gfx-rs/wgpu/blob/94dd3a7dc7f996345261d3aa256dccfc0ac7c49d/naga/src/proc/constant_evaluator.rs#L1719) as `f16::to_u32()` returns None. `f16::to_u32()` under the hood calls `f16::to_f32()` which is infallible, and then calls `f32::to_u32()` (from `num_traits::ToPrimitive`) which if I'm following the code correctly is implemented [here](https://github.com/rust-num/num-traits/blob/022f250f7c12955166dd2ebe776f74e1707312a2/src/cast.rs#L318).

**Repro steps**
```wgsl
enable f16;
fn test() {
  var a = u32(-1h);
}
```

**Expected vs observed behavior**

I would expect this to evaluate to `var a = 0u` as it does for `f32` or `AbstractFloat`.

Instead we get this panic:
```
thread 'main' panicked at naga/src/proc/constant_evaluator.rs:1719:60:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/std/src/panicking.rs:692:5
   1: core::panicking::panic_fmt
             at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/core/src/panicking.rs:75:14
   2: core::panicking::panic
             at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/core/src/panicking.rs:145:5
   3: core::option::unwrap_failed
             at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/core/src/option.rs:2015:5
   4: core::option::Option<T>::unwrap
             at /home/jamie/.rustup/toolchains/1.85-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:978:21
   5: naga::proc::constant_evaluator::ConstantEvaluator::cast
             at ./naga/src/proc/constant_evaluator.rs:1719:44
   6: naga::proc::constant_evaluator::ConstantEvaluator::try_eval_and_append_impl
             at ./naga/src/proc/constant_evaluator.rs:942:36
   7: naga::proc::constant_evaluator::ConstantEvaluator::try_eval_and_append
             at ./naga/src/proc/constant_evaluator.rs:802:35
   8: naga::front::wgsl::lower::ExpressionContext::append_expression
             at ./naga/src/front/wgsl/lower/mod.rs:500:9
   9: naga::front::wgsl::lower::construction::<impl naga::front::wgsl::lower::Lowerer>::construct
             at ./naga/src/front/wgsl/lower/construction.rs:561:20
  10: naga::front::wgsl::lower::Lowerer::expression_for_reference
             at ./naga/src/front/wgsl/lower/mod.rs:2068:30
  11: naga::front::wgsl::lower::Lowerer::expression_for_abstract
             at ./naga/src/front/wgsl/lower/mod.rs:2001:20
  12: naga::front::wgsl::lower::Lowerer::type_and_init
             at ./naga/src/front/wgsl/lower/mod.rs:1292:32
  13: naga::front::wgsl::lower::Lowerer::statement
             at ./naga/src/front/wgsl/lower/mod.rs:1542:45
  14: naga::front::wgsl::lower::Lowerer::block
             at ./naga/src/front/wgsl/lower/mod.rs:1470:13
  15: naga::front::wgsl::lower::Lowerer::function
             at ./naga/src/front/wgsl/lower/mod.rs:1377:24
  16: naga::front::wgsl::lower::Lowerer::lower
             at ./naga/src/front/wgsl/lower/mod.rs:1105:40
  17: naga::front::wgsl::Frontend::inner
             at ./naga/src/front/wgsl/mod.rs:50:22
  18: naga::front::wgsl::Frontend::parse
             at ./naga/src/front/wgsl/mod.rs:44:9
  19: naga::front::wgsl::parse_str
             at ./naga/src/front/wgsl/mod.rs:68:5
  20: naga::parse_input
             at ./naga-cli/src/bin/naga.rs:623:26
  21: naga::run
             at ./naga-cli/src/bin/naga.rs:478:9
  22: naga::main
             at ./naga-cli/src/bin/naga.rs:361:21
  23: core::ops::function::FnOnce::call_once
             at /home/jamie/.rustup/toolchains/1.85-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
```

### PR DESCRIPTION
**Pull request by jamienicol**
This avoids a panic due to f16::to_u32()/f16::to_u64() returning None when the value of the f16 is <= -1.0. The correct behaviour when converting from a floating point to an integer when the value is out of range is to clamp to the nearest value that is representable in both the source and destination type. ie zero for negative numbers.

**Connections**
Fixes #7420 

**Description**
See above, or #7420

**Testing**
Added snapshot tests

**Squash or Rebase?**

Either

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy --tests`. If applicable, add:
- [x] Run `cargo xtask test` to run tests.

### CODE REVIEW

File: naga/tests/in/wgsl/f16.wgsl [hunk 1]
```diff
@@ -61,6 +61,14 @@ fn f16_function(x: f16) -> f16 {
    // Constructing a vec3<i64> from a i64
    val += vec3<f16>(input_uniform.val_f16).z;
 
+   // Cast f16 literals to other types
+   output.val_i32 = i32(65504h);
+   output.val_i32 = i32(-65504h);
+   output.val_u32 = u32(65504h);
+   output.val_u32 = u32(-65504h);
+   output.val_f32 = f32(65504h);
+   output.val_f32 = f32(-65504h);
```
Comment (ErichDonGubler) [inline naga/tests/in/wgsl/f16.wgsl:71]: note: To myself: This value should set all significand bits to 1.

suggestion: Maybe a quick comment on this? 😀

question: Is there any other rationale or explanation for these values that we might usefully add here to this test? Maybe "cross product of min and max values converted to every other scalar type"?

**Review by ErichDonGubler — APPROVED** [review]
LGTM!

File: naga/tests/in/wgsl/f16.wgsl (hunk 1 above)
Comment (jamienicol) [reply naga/tests/in/wgsl/f16.wgsl:71]: It’s chose this more because it’s the max and min finite number, rather than all significand bits being 1 (though of course those two things aren’t just coincidental). Will add a comment.

Converting any negative to unsigned was the only existing bug here, but corner cases such as maxes and mins often make for useful tests. Not really got any more rationale than that
